### PR TITLE
Post visibility better html and a11y

### LINKED ISSUE #1312: PostVisibility component: invalid HTML and better accessibility
When setting the post visibility, a "flying menu/panel" opens with the visibility settings:

screen shot 2017-06-20 at 17 48 41

Some issues noticed so far:

Invalid HTML: label elements can't contain div elements

As per the WordPress [accessibility coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/#labeling), labels should be explicit: that means not wrapping the element and using for/id attributes. Similar issues were already discussed and addressed for other components, so I'm avoiding to repeat the good reasons behind this requirement.

The password protected field has no label. Placeholders should not be used as replacement for labels... see also https://core.trac.wordpress.org/ticket/40331

The toggle button should use an `aria-expanded` attribute to give feedback about the PostVisibility panel opened state.

When closing the panel, focus should be moved back to the toggle button.

Pressing Escape should close the panel.

The radio buttons should be wrapped within a `<fieldset>` element and the `Post Visibility` text should be the fieldset legend: not everything must be a div, please use semantic elements for text when appropriate.

The radio buttons descriptions should be associated to the radio buttons using `aria-describedby` and an ID.

Not sure about Italic text: personally, I found it hard to read especially for text longer then one line. This is more a design decision though.

Not sure about the JS confirm when choosing "Private".

Aside, as mentioned in another issue, the button text (e.g. "Public") should reflect the available action e.g. "Set visibility" and not the option current state. This applies also to the Publish button that shows "Immediately" or the publish date: those are the _state_ of the option, not the action associated with the button.

### PR DESCRIPTION
**Pull request by afercia**
This PR tries to improve the Post Visibility "popover" HTML semantic and accessibility:
- valid HTML
- explicitly associated labels
- added fieldset and legend, see why this is important especially for radio buttons: https://accessibility.blog.gov.uk/2016/07/22/using-the-fieldset-and-legend-elements/
- the password protected field has now a proper label
- changed the password protected placeholder text to be more a "hint": placeholders shouldn't be used as replacement for labels, see https://core.trac.wordpress.org/ticket/40331
- the toggle button now uses an aria-expanded attribute
- don't close the panel when setting visibility to "Private", for consistency with the other 2 options

Not sure why `setPrivate()` uses `onSave()`, would need some feedback here since that seems to trigger a re-rendering of the "popover". /cc @aduth 

Todo:
- when closing the panel, focus should be moved back to the toggle button
- pressing Escape should close the panel (and move focus back)
- when setting visibility to "Private": interaction with the JS confirm: is it really needed?

These points should probably be addressed in #1204 as @aduth mentioned

Note: Needs to be tested in IE/Edge (I suspect the legend element would need some additional styling for those browsers)

Fixes #1312

### CODE REVIEW

**Comment by afercia** [comment]
OK so I've just noticed a (pre-existing) bug:
in Firefox (and maybe other browsers too?) it is not possible to select a different radio button using the down and up arrow keys, note this is the native behavior with radio buttons.
Looks like the radio buttons need a `name` attribute to make Firefox understand they belong to the same group.

**Comment by afercia** [comment]
Screenshot: Firefox and NVDA correctly announcing the fieldset legend, the count of radio buttons, and the descriptions:

screenshot 106

**Comment by aduth** [comment]
>Not sure why `setPrivate()` uses `onSave()`, would need some feedback here since that seems to trigger a re-rendering of the "popover".

I think this was inspired by Calypso's implementation, which prompts the user to immediately publish the post. It was a workaround to an issue caused by Calypso's "noisy" autosave, where drafts are saved very frequently without user intervention (i.e. don't need to press "Save Draft"). It's an issue because you can't save a draft with "Privately Published" as the status (since private and draft are mutually exclusive statuses).

In wp-admin this is pretty unexpected as well:

1. Navigate to Posts > Add New
2. Enter title or content
3. Set Visibility to "Privately Published"
4. Press "Save Draft"

__Expected:__ A draft is saved
__Actual:__ The post is published privately

Depending on the direction we take with #1773, we might need something like this with Gutenberg as well. Or: a way to respect that a user wants the post to be published privately in draft meta. Or: autosave excluding status.

**Comment by aduth** [comment]
>pressing Escape should close the panel (and move focus back)

Should the panel also close when focus leaves (guessing yes)? Example: Press tab while focused in the password input moves to the next Post Settings field, but the panel remains open.

File: editor/sidebar/post-visibility/index.js [hunk 1]
```diff
@@ -56,8 +56,8 @@ class PostVisibility extends Component {
 		const { onUpdateVisibility, onSave } = this.props;
 
 		onUpdateVisibility( 'private' );
+		this.setState( { hasPassword: false } );
```
Comment (aduth) [inline editor/sidebar/post-visibility/index.js]: Outside the scope of this pull request, but I don't like how we have to manually manage password state. Seems like it should be inferred by the `password` prop, or at least consolidated to one of the lifecycle methods (e.g. `componentWillUpdate`). I might plan to tackle this separately.

**Review by aduth — APPROVED** [review]

File: editor/sidebar/post-visibility/index.js [hunk 2]
```diff
@@ -106,35 +106,61 @@ class PostVisibility extends Component {
...
-						) ) }
+						<fieldset>
+							<legend className="editor-post-visibility__dialog-legend">
+								{ __( 'Post Visibility' ) }
+							</legend>
+							{ visibilityOptions.map( ( { value, label, info, onSelect, checked } ) => (
+								<div key={ value } className="editor-post-visibility__choice">
+									<input
+										type="radio"
+										name="editor-post-visibility__setting"
```
Comment (aduth) [inline editor/sidebar/post-visibility/index.js]: It's not likely we'd have more than one of these components on the page, but given that we're injecting `instanceId` anyways, maybe we should include it in the input name?

File: editor/sidebar/post-visibility/index.js (hunk 1 above)
Comment (afercia) [inline editor/sidebar/post-visibility/index.js]: I'd agree it should be addressed separately.

File: editor/sidebar/post-visibility/index.js (hunk 2 above)
Comment (afercia) [inline editor/sidebar/post-visibility/index.js]: Will try.

**Comment by afercia** [comment]
> Should the panel also close when focus leaves (guessing yes)? Example: Press tab while focused in the password input moves to the next Post Settings field, but the panel remains open.

Yeah, I guess it should.

> Depending on the direction we take with #1773, we might need something like this with Gutenberg as well. Or: a way to respect that a user wants the post to be published privately in draft meta. Or: autosave excluding status.

I understand. I can only say that from an usability and accessibility perspective the JS confirm is a terrible experience 🙂  I'd really like to try a different mechanism here, but it makes sense to postpone this issue, pending decisions on autosave.

**Comment by afercia** [comment]
Added `instanceId` to the radio buttons name attribute. Will split closing the panel on Escape/onBlur and moving back the focus in a separate issue (I'd need some feedback and don't want to block this PR).

**Pull request merged**